mettle/worker/worker.go

```diff
@@ -180,21 +181,23 @@ func runForever(instanceName, requestQueue, responseQueue, name, storage, dir, c
 	ch := make(chan os.Signal, 2)
 	signal.Notify(ch, syscall.SIGHUP, syscall.SIGINT, syscall.SIGQUIT, syscall.SIGABRT, syscall.SIGTERM)
 	ctx, cancel := context.WithCancel(context.Background())
+	var wg sync.WaitGroup
+	wg.Add(1)
 	go func() {
+		defer cancel()
+		defer wg.Done()
 		sig := <-ch
 		if immediateShutdown {
 			w.forceShutdown(fmt.Sprintf("Received shutdown signal %s, shutting down...", sig))
+			return
 		}
 		log.Warning("Received signal %s, shutting down when any in-progress task completes or times out", sig)
-		cancel()
-
-		for {
-			select {
-			case <-time.After(5 * time.Minute):
-				w.forceShutdown("Exceeded timeout, shutting down immediately...")
-			case sig = <-ch:
-				w.forceShutdown(fmt.Sprintf("Received another signal %s, shutting down immediately...", sig))
-			}
+
+		select {
+		case <-time.After(5 * time.Minute):
+			w.forceShutdown("Exceeded timeout, shutting down immediately...")
+		case sig = <-ch:
+			w.forceShutdown(fmt.Sprintf("Received another signal %s, shutting down immediately...", sig))
```

Comment:
I know this has not been added in this PR, but we should Nack in the main routine, not in this one. This routine should only cancel the context, so that the main routine knows it should exit. This prevents data races and makes the code easier to understand.